_posts/2023-03-20-leveraging-ffis.md

```diff
@@ -0,0 +1,44 @@
+
+While leveraging FFI to unify our codebases is still a work in progress, we have had our FFI based Asherah libraries in production successfully for months.  Platform support for FFI has improved greatly across the board over the years.  For example, we've noticed that OpenJDK has a new FFI interface in incubation that will facilitate the consumption of C shared libraries rather than using Java Native Access.
+
+## Go vs Rust
+We're now working on a Rust based shared library that will provide shared configuration management.  We've found there are advantages and disadvantages to using Rust as the base language for the shared libraries versus Go.  With Go, we ran into a significant issue where Go does not play nicely with Alpine's use of musl libc.  Go's compiler only produces C shared libraries with the `init-exec` thread local storage model.  However, musl libc does not support using `dlopen()` to dynamically load libraries with the `init-exec` model, as this would require space to be reserved at initial load time.  If the Go compiler were able to produce shared libraries with the `global-dynamic` thread local storage model like gcc does, there wouldn't be a compatibility issue.  The Go standard library also makes glibc assumptions like the idea that shared libraries would be passed argc/argv despite incompatibility with the ELF standard, and crashes if those parameters are not passed.  Unfortunately, this means that any of our systems consuming Go based shared libraries on [Alpine](https://www.alpinelinux.org/), are forced to switch to a glibc based Linux distribution like Debian.  We hope to see these compatibility issues addressed in the future as the Go compiler evolves to better work with standardized systems and not presuming glibc behaviors.
+
+## Conclusion
+Given our results so far, we feel that despite the bumps along the way, our strategy of using a shared code base written in Rust or Go to develop libraries to be consumed by our other platforms is paying off and will continue to pay off, versus the strategy of implementing the same functionality repeatedly in each supported language.
+
+If helping solve the challenge of shared libraries interests you, check out our [current job openings](https://careers.godaddy.com/search-jobs).
+
+_*) Cover Photo Attribution: Photo by <a href="https://www.flickr.com/photos/nasawebbtelescope/52692145572/in/album-72177720305127361/">NASA’s Webb Reveals Intricate Networks of Gas and Dust in Nearby Galaxies</a> on <a href="https://www.flickr.com/photos/nasawebbtelescope/52692145572/in/album-72177720305127361/">NASA</a>_
```

Comment:
What styling are we trying to use here? it renders weird... assuming we would just want this to italicized text?
<img width="1182" alt="Screen Shot 2023-03-20 at 2 56 19 PM" src="https://user-images.githubusercontent.com/63810222/226474425-9c9e6f14-5d51-4438-8164-39c4af4d7b1c.png">

Reply:
```suggestion
*Cover Photo Attribution: Photo by <a href="https://www.flickr.com/photos/nasawebbtelescope/52692145572/in/album-72177720305127361/">NASA’s Webb Reveals Intricate Networks of Gas and Dust in Nearby Galaxies</a> on <a href="https://www.flickr.com/photos/nasawebbtelescope/52692145572/in/album-72177720305127361/">NASA</a>*
```

Reply:
It's same on the other blog posts, [example](https://github.com/godaddy/engineering/pull/236/files#diff-8e27f8fccf8c0bff3b02c2a7b702cf07fb44af0e380e968356a4954eeab3618aR141). It's basically:

```
<i>*) Cover Photo Attribution: ...</i>
```

Reply:
oh that '*)' is weird - looks like a mistake tbh. but otherwise it's good, so I'll approve